### PR TITLE
[SID-1688] Add support for the username handle type

### PR DESCRIPTION
**Pull request by ikovic**
## Description

[YouTrack issue](https://todo.irdcorp.dev/issue/SID-1688)

Support the `username` handle type. This feature is entirely opt in - it can be enabled by passing `username` to the list of `allowedHandleTypes` for the given factor. Currently only `password` can work with usernames, but later `totp` will work as well.

## Testing

`slashid/react@1.24.1-username.1` is the latest beta, containing all the username related changes.

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have generated the new version of the docs website and smoke tested it
- [ ] I have checked that my changes haven't caused semantic errors in the existing docs
- [ ] I have generated a `changeset` if my change affects the published packages

### CODE REVIEW

File: packages/react/src/domain/handles.ts [hunk 1]
```diff
@@ -58,6 +61,13 @@ function getPossibleHandleTypes(
     handleTypes.add("phone_number");
   }
 
+  if (
+    FACTORS_WITH_USERNAME.includes(factor.method) &&
+    isHandleTypeAllowed(factor, "username")
```
Comment (kaspermroz) [inline packages/react/src/domain/handles.ts:66]: I feel like we could combine these two in one fn for simplicity

File: packages/react/src/components/dynamic-flow/configured-handle-form.tsx [hunk 2]
```diff
@@ -15,11 +15,13 @@ import {
 export const TAB_NAME = {
   email: "email",
   phone: "phone",
+  username: "username",
```
Comment (kaspermroz) [inline packages/react/src/components/dynamic-flow/configured-handle-form.tsx:18]: Does it make sense to stick to tabs here? Won't this be a little bit too cramped?

File: packages/react/src/components/dynamic-flow/handle-form.tsx [hunk 3]
```diff
@@ -105,6 +106,27 @@ export const HandleForm: React.FC<Props> = ({
...
+          id={`sid-input-${handleType}`}
+          name={handleType}
+          label={text["initial.handle.username"]}
+          placeholder={text["initial.handle.username.placeholder"]}
+          value={values[handleType] ?? ""}
+          onChange={registerField(handleType, {
+            defaultValue,
+            validator: (value) => {
+              if (!isValidUsername(value)) {
+                return { message: text["validationError.email"] };
```
Comment (kaspermroz) [inline packages/react/src/components/dynamic-flow/handle-form.tsx:122]: Should probably use a different validation error

File: packages/react/src/components/form/initial/controls.tsx [hunk 4]
```diff
@@ -46,11 +50,13 @@ export const FACTOR_LABEL_MAP: Record<
 export const TAB_NAME = {
   email: "email",
   phone: "phone",
+  username: "username",
 };
 
 export const tabIDByHandle: Record<HandleType, string> = {
   phone_number: TAB_NAME.phone,
   email_address: TAB_NAME.email,
+  username: TAB_NAME.username,
```
Comment (kaspermroz) [inline packages/react/src/components/form/initial/controls.tsx:59]: Same concern as for dynamic flow
> Does it make sense to stick to tabs here? Won't this be a little bit too cramped?

File: packages/react/src/components/form/initial/controls.tsx [hunk 5]
```diff
@@ -363,6 +396,27 @@ const HandleInput: React.FC<PropsInternal> = ({
...
+          id={`sid-input-${handleType}`}
+          name={handleType}
+          label={text["initial.handle.username"]}
+          placeholder={text["initial.handle.username.placeholder"]}
+          value={values[handleType] ?? ""}
+          onChange={registerField(handleType, {
+            defaultValue,
+            validator: (value) => {
+              if (!isValidUsername(value)) {
+                return { message: text["validationError.email"] };
```
Comment (kaspermroz) [inline packages/react/src/components/form/initial/controls.tsx:412]: Same concern as for dynamic flow
> Should probably use a different validation error

File: packages/react/src/components/form/form-username.test.tsx [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        </ConfigurationProvider>
+      </TestSlashIDProvider>
+    );
+
+    expect(
+      screen.queryByTestId("sid-tabs-trigger-username")
+    ).not.toBeInTheDocument();
+  });
+
+  test("username tab should not render unless explicitly configured using allowedHandleTypes", () => {
```
Comment (kaspermroz) [inline packages/react/src/components/form/form-username.test.tsx:35]: "username tab should render when explicitly configured using allowedHandleTypes"

File: packages/react/src/domain/handles.ts (hunk 1 above)
Comment (ikovic) [reply packages/react/src/domain/handles.ts:68]: I understand the concern, right now I don't see an obvious and easy way of making this more concise so I'd rather skip this in the scope of this PR. Happy to go back and refactor once there's a clear idea on how to do it exactly

File: packages/react/src/components/dynamic-flow/configured-handle-form.tsx (hunk 2 above)
Comment (ikovic) [reply packages/react/src/components/dynamic-flow/configured-handle-form.tsx:18]: I'd leave this open to feedback as usernames are opt in and we don't expect customers to run all three possible handle types at the same time.

File: packages/react/src/components/dynamic-flow/handle-form.tsx (hunk 3 above)
Comment (ikovic) [reply packages/react/src/components/dynamic-flow/handle-form.tsx:122]: sure, thanks!

Comment (ikovic) [reply packages/react/src/components/dynamic-flow/handle-form.tsx:122]: fixed

File: packages/react/src/components/form/initial/controls.tsx (hunk 5 above)
Comment (ikovic) [reply packages/react/src/components/form/initial/controls.tsx:412]: fixed

File: packages/react/src/components/form/initial/controls.tsx (hunk 4 above)
Comment (ikovic) [reply packages/react/src/components/form/initial/controls.tsx:59]: https://github.com/slashid/javascript/pull/248#discussion_r1590882155

File: packages/react/src/components/form/form-username.test.tsx (hunk 6 above)
Comment (ikovic) [reply packages/react/src/components/form/form-username.test.tsx]: thanks, fixed!

**Review by kaspermroz — APPROVED** [review]

File: packages/react/src/components/text/constants.ts [hunk 7]
```diff
@@ -11,8 +11,10 @@ export const TEXT = {
   "initial.authenticationMethod": "Authentication method",
   "initial.handle.email": "Email address",
   "initial.handle.phone": "Phone number",
```
Comment (matilde-slashid) [inline packages/react/src/components/text/constants.ts:13]:
```suggestion
  "initial.handle.phone": "Phone",
```

Comment (matilde-slashid) [reply packages/react/src/components/text/constants.ts:13]: like this they should look better if we have 3 side by side

File: packages/react/src/components/text/constants.ts [hunk 8]
```diff
@@ -118,9 +120,18 @@ export const TEXT = {
...
     "There has been an error while submitting your form. Please try again.",
+  "error.title.rateLimit": "Too many attempts...",
   "error.subtitle.rateLimit":
     "Your request has been rate limited. Please try again later.",
+  "error.title.recoverNonReachableHandleType": "Cannot reset password",
+  "error.subtitle.recoverNonReachableHandleType":
+    "Please use an email address or a phone number to reset your password.",
+  "error.title.noPasswordSet": "No password set",
+  "error.subtitle.noPasswordSet":
+    "Please contact support to set a password or log in with a phone number or an email address.",
```
Comment (matilde-slashid) [inline packages/react/src/components/text/constants.ts:131]:
```suggestion
    "Contact support to set a password or log in with your phone number or email.",
```

**Pull request merged**